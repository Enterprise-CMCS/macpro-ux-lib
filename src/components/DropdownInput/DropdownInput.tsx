import { Icon } from "components/Icon/Icon";
import React, {
  Dispatch,
  SetStateAction,
  useEffect,
  useRef,
  useState,
} from "react";

type IntrinsicElements = JSX.IntrinsicElements["select"];

interface DropdownData {
  value: string | number;
  displayString: string;
}

interface DropdownProps extends IntrinsicElements {
  data: DropdownData[];
  label: string;
  readOnly?: boolean;
  setValue?: Dispatch<SetStateAction<any>>;
  value?: string | number | undefined;
  select: React.ReactNode;
}

/**
 * Dropdown component
 *
 * This component is a wrapper around the HTML `<select>` element and provides the basic functionality of a dropdown along with accessibility and styling.
 * A key enhancement to the standard dropdown is the ability for a user to type to search for an option.
 *
 * @param {string}        [className]    A class name that will be applied on the select element.
 * @param {DropdownData}  data           Data used to populate the dropdown.
 * @param {string}        id             The id of the dropdown.
 * @param {string}        label          String used to label the dropdown in the UI.
 * @param {string}        name           Name of the dropdown used to identify it in the context of a form.
 * @param {string}        [placeholder]  Placeholder text to be displayed in the input.
 * @param {boolean}       readOnly       Sets input field to read-only. Effectively disables type-ahead search.
 */
export const DropdownInput: React.FC<DropdownProps> = ({
  className,
  data,
  id,
  label,
  name,
  placeholder,
  readOnly = false,
  select,
  setValue,
  value,
  ...rest
}) => {
  const [activeDescendant, setActiveDescendant] = useState<string>("");
  const [dropdownData, setDropdownData] = useState<DropdownData[]>(data);
  const [hidden, setHidden] = useState<boolean>(true);
  const [inputValue, setInputValue] = useState<string>("");

  if (value === undefined && setValue === undefined)
    [value, setValue] = useState<string | number | undefined>("");

  const closeDropdown = () => {
    setHidden(true);
    setActiveDescendant("");
  };

  /*
   * Handles ArrowDown when input has focus.
   *
   * If a value was previously selected, focus should shift to that element on ArrowDown.
   * If a value has not yet been selected, focus goes to the first option in the list.
   */
  const handleInputArrowDown = () => {
    const selectedElement = document.getElementsByClassName(
      "usa-combo-box__list-option--selected"
    )[0] as HTMLElement;

    const firstElement = document.getElementsByClassName(
      "usa-combo-box__list-option"
    )[0] as HTMLElement;

    const elem = selectedElement ?? firstElement;

    if (elem) {
      elem.focus();
      setActiveDescendant(elem.id);
    }
  };

  // Collapse dropdown onBlur – do not collapse if focus shifts to a child element
  const handleBlur = (e: React.FocusEvent) => {
    if (!hidden) {
      const currentTarget = e.currentTarget;
      requestAnimationFrame(() => {
        if (!currentTarget.contains(document.activeElement)) {
          closeDropdown();
        }
      });
    }
  };

  // triggered by click or click-equivalent event (Enter/Tab)
  const handleItemClick = (itemValue: string | number) => {
    setValue && setValue(itemValue);
    closeDropdown();
  };

  // Sets inputValue text and filters dropdown options
  const handleInput = (inputValue: string) => {
    setInputValue(inputValue);
    setDropdownData(
      data.filter((itm) => {
        const displayStr = itm.displayString.toLowerCase();
        const searchString = inputValue.toLowerCase();
        return displayStr.includes(searchString);
      })
    );
  };

  const getDisplayString = () => {
    const obj = data.find((itm) => itm.value === value);
    return obj?.displayString ?? "";
  };

  // when value updates, inputValue should match value
  useEffect(() => {
    setInputValue(getDisplayString());
    setDropdownData(data);
  }, [value, data]);

  return (
    <>
      <label className="usa-label" htmlFor={id} id={`${id}-label`}>
        {label}
      </label>
      <div
        className={`usa-combo-box${
          value || inputValue ? " usa-combo-box--pristine" : ""
        }`}
        data-enhanced="true"
        onBlur={(e) => handleBlur(e)}
      >
        {select}
        <input
          aria-activedescendant={activeDescendant}
          aria-autocomplete="list"
          aria-controls={`${id}--list`}
          aria-describedby={`${id}--assistiveHint`}
          aria-expanded="false"
          aria-owns={`${id}--list`}
          autoCapitalize="none"
          autoComplete="off"
          className="usa-combo-box__input"
          id={id}
          onChange={(e) => handleInput(e.target.value)}
          onFocus={() => setHidden(false)}
          onKeyDown={(e) => {
            if (e.key === "ArrowDown") {
              e.preventDefault();
              handleInputArrowDown();
            }
          }}
          placeholder={placeholder}
          type="text"
          readOnly={readOnly}
          role="combobox"
          value={inputValue}
        />
        <span className="usa-combo-box__clear-input__wrapper" tabIndex={-1}>
          <button
            hidden={!value}
            type="button"
            className="usa-combo-box__clear-input"
            aria-label="Clear the select contents"
            onClick={() => {
              setValue && setValue("");
            }}
          >
            <Icon name="close" />
          </button>
        </span>
        <span className="usa-combo-box__input-button-separator">&nbsp;</span>
        <span className="usa-combo-box__toggle-list__wrapper" tabIndex={-1}>
          <button
            type="button"
            tabIndex={-1}
            className="usa-combo-box__toggle-list"
            aria-label="Toggle the dropdown list"
            onClick={() => setHidden(!hidden)}
          >
            <Icon name="expand_more" />
          </button>
        </span>
        <ul
          tabIndex={-1}
          id={`${id}--list`}
          className="usa-combo-box__list"
          role="listbox"
          aria-labelledby={`${id}-label`}
          hidden={hidden}
        >
          {dropdownData.map((item, index) => {
            const itemId = `${id}--option-${index}`;
            return (
              <DropdownItem
                {...{
                  activeDescendant,
                  data,
                  handleItemClick,
                  index,
                  item,
                  itemId,
                  key: itemId,
                  setActiveDescendant,
                  value,
                }}
              />
            );
          })}
        </ul>
        <div className="usa-combo-box__status usa-sr-only" role="status">
          {!hidden && `${data.length} results available.`}
        </div>
        <span id={`${id}--assistiveHint`} className="usa-sr-only">
          When autocomplete results are available use up and down arrows to
          review and enter to select. Touch device users, explore by touch or
          with swipe gestures.
        </span>
      </div>
    </>
  );
};

/*
 * UseFocus takes a ref and shifts focus to the referenced element.
 * Used here as a way to shift focus in a self-referential way within DropdownItem.
 */
const UseFocus = (): [React.MutableRefObject<any>, () => void] => {
  const htmlElRef = useRef<any>(null);
  const setFocus = () => {
    htmlElRef.current.focus();
  };

  return [htmlElRef, setFocus];
};

interface DropdownItemProps {
  activeDescendant: string;
  data: DropdownData[];
  handleItemClick: (arg: string | number) => void;
  index: number;
  item: DropdownData;
  itemId: string;
  setActiveDescendant: any;
  value: string | number | undefined;
}

/*
 * DropdownItem component is a helper component for Dropdown.
 * Represents each item in the dropdown menu and contains much of the functionality and logic.
 */
const DropdownItem = ({
  activeDescendant,
  data,
  handleItemClick,
  index,
  item,
  itemId,
  setActiveDescendant,
  value,
}: DropdownItemProps) => {
  const [ref, setFocus] = UseFocus();
  const selected = item.value === value;

  let focused = false;
  if (itemId === activeDescendant) focused = true;
  else if (!activeDescendant && selected) focused = true;
  else if (!activeDescendant && !value && index === 0) focused = true;

  let tabIndex = -1;
  if (activeDescendant && focused) tabIndex = 0;
  else if (!activeDescendant && index === 0) tabIndex = 0;

  // takes a string representing arrow key press and shifts focus down or up accordingly
  const handleArrow = (arrow: "ArrowDown" | "ArrowUp") => {
    const currentElement = document.getElementsByClassName(
      "usa-combo-box__list-option--focused"
    )[0];

    let elemToFocus;
    if (arrow === "ArrowDown") {
      elemToFocus = currentElement?.nextElementSibling as HTMLElement;
    } else if (arrow === "ArrowUp") {
      elemToFocus = currentElement?.previousElementSibling as HTMLElement;
    }

    if (elemToFocus) {
      elemToFocus.focus();
      setActiveDescendant(elemToFocus.id);
    }
  };

  /*
   * disable default behavior
   * on Enter/Tab set value to be focused element value
   * on ArrowDown/ArrowUp shift focus appropriately
   */
  const handleKeyDown = (e: React.KeyboardEvent<HTMLLIElement>) => {
    e.preventDefault();

    switch (e.key) {
      case "Tab":
      case "Enter":
        handleItemClick(item.value);
        break;
      case "ArrowDown":
      case "ArrowUp":
        handleArrow(e.key);
        break;
    }
  };

  return (
    <li
      aria-setsize={data.length}
      aria-posinset={index + 1}
      aria-selected="false"
      className={`usa-combo-box__list-option${
        focused ? " usa-combo-box__list-option--focused" : ""
      }${selected ? " usa-combo-box__list-option--selected" : ""}`}
      data-value={item.value}
      id={itemId}
      ref={ref}
      tabIndex={tabIndex}
      role="option"
      onClick={() => handleItemClick(item.value)}
      onKeyDown={(e) => handleKeyDown(e)}
      onMouseOver={() => {
        setActiveDescendant(itemId);
        setFocus();
      }}
    >
      {item.displayString}
    </li>
  );
};
