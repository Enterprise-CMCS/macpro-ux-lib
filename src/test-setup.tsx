import React, { ReactElement } from "react";
import { render, RenderOptions } from "@testing-library/react";
import "@testing-library/jest-dom";

// Here we can set all providers for themes, stores and other stuff
const AllTheProviders = ({ children }: { children: JSX.Element }) => {
  return <>{children}</>;
};

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, "wrapper">
) => render(ui, { wrapper: AllTheProviders, ...options });

export * from "@testing-library/react";
export { customRender as render };
