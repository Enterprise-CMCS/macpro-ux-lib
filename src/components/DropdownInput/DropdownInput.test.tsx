import React from "react";
import { fireEvent as userEvent, render } from "../../test-setup";
import fireEvent from "@testing-library/user-event";

import { DropdownInput } from "./DropdownInput";
import data from "./data.json";

describe("Test the DropdownInput component", () => {
  let container: HTMLElement;
  let inputElem: Element;
  let listElem: Element;

  beforeEach(() => {
    container = render(
      <DropdownInput
        data={data}
        id="fruit"
        label="Select a fruit"
        name="fruit-dropdown"
      />
    ).container;
    inputElem = container.getElementsByClassName("usa-combo-box__input")[0];
    listElem = container.getElementsByClassName("usa-combo-box__list")[0];

    jest
      .spyOn(window, "requestAnimationFrame")
      .mockImplementation((cb: any) => cb());
  });

  it("should render correctly", () => {
    expect(inputElem).toBeInTheDocument();
    expect(listElem).toBeInTheDocument();

    // list should exist but not be visible
    expect(listElem.children.length).toBe(data.length);
    expect(listElem).not.toBeVisible();
  });

  it("should expand and collapse", () => {
    // should render collapsed
    expect(listElem).not.toBeVisible();

    // expand on input focus
    userEvent.focus(inputElem);
    expect(listElem).toBeVisible();

    // collapse on toggle click
    const toggleArrow = container.getElementsByClassName(
      "usa-combo-box__toggle-list"
    )[0];
    userEvent.click(toggleArrow);
    expect(listElem).not.toBeVisible();

    // expand on input focus
    userEvent.focus(inputElem);
    expect(listElem).toBeVisible();

    // collapse on input blur
    userEvent.blur(inputElem);
    expect(listElem).not.toBeVisible();
  });

  it("should be able to select elements", () => {
    userEvent.focus(inputElem);

    // select an option
    const apple = container.getElementsByClassName(
      "usa-combo-box__list-option"
    )[0];
    userEvent.click(apple);

    // should contain the data
    expect(inputElem).toHaveDisplayValue("Apple");
  });

  it("should be able to clear a selection", () => {
    const clearBtn = container.getElementsByClassName(
      "usa-combo-box__clear-input"
    )[0];
    expect(clearBtn).not.toBeVisible();

    userEvent.focus(inputElem);
    const apple = container.getElementsByClassName(
      "usa-combo-box__list-option"
    )[0];
    userEvent.click(apple);

    // should contain the data
    expect(inputElem).toHaveDisplayValue("Apple");
    expect(clearBtn).toBeVisible();

    userEvent.click(clearBtn);

    // should be empty
    expect(inputElem).toHaveDisplayValue("");
    expect(clearBtn).not.toBeVisible();
  });

  it("while input is focused, arrow down should focus on first item", () => {
    userEvent.focus(inputElem);
    userEvent.keyDown(inputElem, { key: "ArrowDown" });

    const apple = container.getElementsByClassName(
      "usa-combo-box__list-option"
    )[0];

    expect(apple).toHaveFocus();
  });

  it("arrows should navigate dropdown list", () => {
    userEvent.focus(inputElem);
    userEvent.keyDown(inputElem, { key: "ArrowDown" });

    const apple = container.getElementsByClassName(
      "usa-combo-box__list-option"
    )[0];
    const apricot = container.getElementsByClassName(
      "usa-combo-box__list-option"
    )[1];

    expect(apple).toHaveFocus();

    // Move down
    userEvent.keyDown(apple, { key: "ArrowDown" });

    expect(apple).not.toHaveFocus();
    expect(apricot).toHaveFocus();

    // Move up
    userEvent.keyDown(apricot, { key: "ArrowUp" });

    expect(apple).toHaveFocus();
    expect(apricot).not.toHaveFocus();

    // Make selection
    userEvent.keyDown(apple, { key: "Enter" });

    expect(inputElem).toHaveDisplayValue("Apple");
  });

  it("selector should jump to current dropdown value on ArrowDown", () => {
    const apple = container.getElementsByClassName(
      "usa-combo-box__list-option"
    )[0];
    const apricot = container.getElementsByClassName(
      "usa-combo-box__list-option"
    )[1];

    // Move down to apricot
    userEvent.focus(inputElem);
    userEvent.keyDown(inputElem, { key: "ArrowDown" });
    userEvent.keyDown(apple, { key: "ArrowDown" });
    expect(apricot).toHaveFocus();

    // collapse list
    userEvent.blur(inputElem);
    expect(listElem).not.toBeVisible();

    // refocus input; ArrowDown should jump to current value
    userEvent.focus(inputElem);
    userEvent.keyDown(window, { key: "ArrowDown" });
    expect(apricot).toHaveFocus();
  });

  it("typing in input should filter dropdown list", () => {
    userEvent.focus(inputElem);
    fireEvent.type(inputElem, "apple");

    expect(listElem.children.length).toBe(4);
  });

  it("hover should focus list item", () => {
    userEvent.focus(inputElem);
    const apple = container.getElementsByClassName(
      "usa-combo-box__list-option"
    )[0];
    const apricot = container.getElementsByClassName(
      "usa-combo-box__list-option"
    )[1];

    userEvent.mouseOver(apple);

    expect(apple).toHaveFocus();
    expect(apricot).not.toHaveFocus();

    userEvent.mouseOver(apricot);

    expect(apple).not.toHaveFocus();
    expect(apricot).toHaveFocus();
  });

  it("component snapshot", () => {
    expect(container).toMatchSnapshot();
  });
});
