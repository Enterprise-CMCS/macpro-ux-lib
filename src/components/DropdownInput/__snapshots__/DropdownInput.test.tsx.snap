// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test the DropdownInput component component snapshot 1`] = `
<div>
  <label
    class="usa-label"
    for="fruit"
    id="fruit-label"
  >
    Select a fruit
  </label>
  <div
    class="usa-combo-box"
    data-enhanced="true"
  >
    <input
      aria-activedescendant=""
      aria-autocomplete="list"
      aria-controls="fruit--list"
      aria-describedby="fruit--assistiveHint"
      aria-expanded="false"
      aria-owns="fruit--list"
      autocapitalize="none"
      autocomplete="off"
      class="usa-combo-box__input"
      id="fruit"
      role="combobox"
      type="text"
      value=""
    />
    <span
      class="usa-combo-box__clear-input__wrapper"
      tabindex="-1"
    >
      <button
        aria-label="Clear the select contents"
        class="usa-combo-box__clear-input"
        hidden=""
        type="button"
      >
        <svg
          aria-hidden="true"
          aria-label="close icon"
          class="usa-icon--size-3"
          color="black"
          fill="currentColor"
          role="img"
        >
          <use
            href="[object Object]#close"
          />
        </svg>
      </button>
    </span>
    <span
      class="usa-combo-box__input-button-separator"
    >
      Â 
    </span>
    <span
      class="usa-combo-box__toggle-list__wrapper"
      tabindex="-1"
    >
      <button
        aria-label="Toggle the dropdown list"
        class="usa-combo-box__toggle-list"
        tabindex="-1"
        type="button"
      >
        <svg
          aria-hidden="true"
          aria-label="expand more icon"
          class="usa-icon--size-3"
          color="black"
          fill="currentColor"
          role="img"
        >
          <use
            href="[object Object]#expand_more"
          />
        </svg>
      </button>
    </span>
    <ul
      aria-labelledby="fruit-label"
      class="usa-combo-box__list"
      hidden=""
      id="fruit--list"
      role="listbox"
      tabindex="-1"
    >
      <li
        aria-posinset="1"
        aria-selected="false"
        aria-setsize="64"
        class="usa-combo-box__list-option usa-combo-box__list-option--focused"
        data-value="apple"
        id="fruit--option-0"
        role="option"
        tabindex="0"
      >
        Apple
      </li>
      <li
        aria-posinset="2"
        aria-selected="false"
        aria-setsize="64"
        class="usa-combo-box__list-option"
        data-value="apricot"
        id="fruit--option-1"
        role="option"
        tabindex="-1"
      >
        Apricot
      </li>
      <li
        aria-posinset="3"
        aria-selected="false"
        aria-setsize="64"
        class="usa-combo-box__list-option"
        data-value="avocado"
        id="fruit--option-2"
        role="option"
        tabindex="-1"
      >
        Avocado
      </li>
      <li
        aria-posinset="4"
        aria-selected="false"
        aria-setsize="64"
        class="usa-combo-box__list-option"
        data-value="banana"
        id="fruit--option-3"
        role="option"
        tabindex="-1"
      >
        Banana
      </li>
      <li
        aria-posinset="5"
        aria-selected="false"
        aria-setsize="64"
        class="usa-combo-box__list-option"
        data-value="blackberry"
        id="fruit--option-4"
        role="option"
        tabindex="-1"
      >
        Blackberry
      </li>
      <li
        aria-posinset="6"
        aria-selected="false"
        aria-setsize="64"
        class="usa-combo-box__list-option"
        data-value="blood orange"
        id="fruit--option-5"
        role="option"
        tabindex="-1"
      >
        Blood orange
      </li>
      <li
        aria-posinset="7"
        aria-selected="false"
        aria-setsize="64"
        class="usa-combo-box__list-option"
        data-value="blueberry"
        id="fruit--option-6"
        role="option"
        tabindex="-1"
      >
        Blueberry
      </li>
      <li
        aria-posinset="8"
        aria-selected="false"
        aria-setsize="64"
        class="usa-combo-box__list-option"
        data-value="boysenberry"
        id="fruit--option-7"
        role="option"
        tabindex="-1"
      >
        Boysenberry
      </li>
      <li
        aria-posinset="9"
        aria-selected="false"
        aria-setsize="64"
        class="usa-combo-box__list-option"
        data-value="breadfruit"
        id="fruit--option-8"
        role="option"
        tabindex="-1"
      >
        Breadfruit
      </li>
      <li
        aria-posinset="10"
        aria-selected="false"
        aria-setsize="64"
        class="usa-combo-box__list-option"
        data-value="buddhas hand citron"
        id="fruit--option-9"
        role="option"
        tabindex="-1"
      >
        Buddha's hand citron
      </li>
      <li
        aria-posinset="11"
        aria-selected="false"
        aria-setsize="64"
        class="usa-combo-box__list-option"
        data-value="cantaloupe"
        id="fruit--option-10"
        role="option"
        tabindex="-1"
      >
        Cantaloupe
      </li>
      <li
        aria-posinset="12"
        aria-selected="false"
        aria-setsize="64"
        class="usa-combo-box__list-option"
        data-value="clementine"
        id="fruit--option-11"
        role="option"
        tabindex="-1"
      >
        Clementine
      </li>
      <li
        aria-posinset="13"
        aria-selected="false"
        aria-setsize="64"
        class="usa-combo-box__list-option"
        data-value="crab apple"
        id="fruit--option-12"
        role="option"
        tabindex="-1"
      >
        Crab apple
      </li>
      <li
        aria-posinset="14"
        aria-selected="false"
        aria-setsize="64"
        class="usa-combo-box__list-option"
        data-value="currant"
        id="fruit--option-13"
        role="option"
        tabindex="-1"
      >
        Currant
      </li>
      <li
        aria-posinset="15"
        aria-selected="false"
        aria-setsize="64"
        class="usa-combo-box__list-option"
        data-value="cherry"
        id="fruit--option-14"
        role="option"
        tabindex="-1"
      >
        Cherry
      </li>
      <li
        aria-posinset="16"
        aria-selected="false"
        aria-setsize="64"
        class="usa-combo-box__list-option"
        data-value="custard apple"
        id="fruit--option-15"
        role="option"
        tabindex="-1"
      >
        Custard apple
      </li>
      <li
        aria-posinset="17"
        aria-selected="false"
        aria-setsize="64"
        class="usa-combo-box__list-option"
        data-value="coconut"
        id="fruit--option-16"
        role="option"
        tabindex="-1"
      >
        Coconut
      </li>
      <li
        aria-posinset="18"
        aria-selected="false"
        aria-setsize="64"
        class="usa-combo-box__list-option"
        data-value="cranberry"
        id="fruit--option-17"
        role="option"
        tabindex="-1"
      >
        Cranberry
      </li>
      <li
        aria-posinset="19"
        aria-selected="false"
        aria-setsize="64"
        class="usa-combo-box__list-option"
        data-value="date"
        id="fruit--option-18"
        role="option"
        tabindex="-1"
      >
        Date
      </li>
      <li
        aria-posinset="20"
        aria-selected="false"
        aria-setsize="64"
        class="usa-combo-box__list-option"
        data-value="dragonfruit"
        id="fruit--option-19"
        role="option"
        tabindex="-1"
      >
        Dragonfruit
      </li>
      <li
        aria-posinset="21"
        aria-selected="false"
        aria-setsize="64"
        class="usa-combo-box__list-option"
        data-value="durian"
        id="fruit--option-20"
        role="option"
        tabindex="-1"
      >
        Durian
      </li>
      <li
        aria-posinset="22"
        aria-selected="false"
        aria-setsize="64"
        class="usa-combo-box__list-option"
        data-value="elderberry"
        id="fruit--option-21"
        role="option"
        tabindex="-1"
      >
        Elderberry
      </li>
      <li
        aria-posinset="23"
        aria-selected="false"
        aria-setsize="64"
        class="usa-combo-box__list-option"
        data-value="fig"
        id="fruit--option-22"
        role="option"
        tabindex="-1"
      >
        Fig
      </li>
      <li
        aria-posinset="24"
        aria-selected="false"
        aria-setsize="64"
        class="usa-combo-box__list-option"
        data-value="gooseberry"
        id="fruit--option-23"
        role="option"
        tabindex="-1"
      >
        Gooseberry
      </li>
      <li
        aria-posinset="25"
        aria-selected="false"
        aria-setsize="64"
        class="usa-combo-box__list-option"
        data-value="grape"
        id="fruit--option-24"
        role="option"
        tabindex="-1"
      >
        Grape
      </li>
      <li
        aria-posinset="26"
        aria-selected="false"
        aria-setsize="64"
        class="usa-combo-box__list-option"
        data-value="grapefruit"
        id="fruit--option-25"
        role="option"
        tabindex="-1"
      >
        Grapefruit
      </li>
      <li
        aria-posinset="27"
        aria-selected="false"
        aria-setsize="64"
        class="usa-combo-box__list-option"
        data-value="guava"
        id="fruit--option-26"
        role="option"
        tabindex="-1"
      >
        Guava
      </li>
      <li
        aria-posinset="28"
        aria-selected="false"
        aria-setsize="64"
        class="usa-combo-box__list-option"
        data-value="honeydew melon"
        id="fruit--option-27"
        role="option"
        tabindex="-1"
      >
        Honeydew melon
      </li>
      <li
        aria-posinset="29"
        aria-selected="false"
        aria-setsize="64"
        class="usa-combo-box__list-option"
        data-value="jackfruit"
        id="fruit--option-28"
        role="option"
        tabindex="-1"
      >
        Jackfruit
      </li>
      <li
        aria-posinset="30"
        aria-selected="false"
        aria-setsize="64"
        class="usa-combo-box__list-option"
        data-value="kiwifruit"
        id="fruit--option-29"
        role="option"
        tabindex="-1"
      >
        Kiwifruit
      </li>
      <li
        aria-posinset="31"
        aria-selected="false"
        aria-setsize="64"
        class="usa-combo-box__list-option"
        data-value="kumquat"
        id="fruit--option-30"
        role="option"
        tabindex="-1"
      >
        Kumquat
      </li>
      <li
        aria-posinset="32"
        aria-selected="false"
        aria-setsize="64"
        class="usa-combo-box__list-option"
        data-value="lemon"
        id="fruit--option-31"
        role="option"
        tabindex="-1"
      >
        Lemon
      </li>
      <li
        aria-posinset="33"
        aria-selected="false"
        aria-setsize="64"
        class="usa-combo-box__list-option"
        data-value="lime"
        id="fruit--option-32"
        role="option"
        tabindex="-1"
      >
        Lime
      </li>
      <li
        aria-posinset="34"
        aria-selected="false"
        aria-setsize="64"
        class="usa-combo-box__list-option"
        data-value="lychee"
        id="fruit--option-33"
        role="option"
        tabindex="-1"
      >
        Lychee
      </li>
      <li
        aria-posinset="35"
        aria-selected="false"
        aria-setsize="64"
        class="usa-combo-box__list-option"
        data-value="mandarine"
        id="fruit--option-34"
        role="option"
        tabindex="-1"
      >
        Mandarine
      </li>
      <li
        aria-posinset="36"
        aria-selected="false"
        aria-setsize="64"
        class="usa-combo-box__list-option"
        data-value="mango"
        id="fruit--option-35"
        role="option"
        tabindex="-1"
      >
        Mango
      </li>
      <li
        aria-posinset="37"
        aria-selected="false"
        aria-setsize="64"
        class="usa-combo-box__list-option"
        data-value="mangosteen"
        id="fruit--option-36"
        role="option"
        tabindex="-1"
      >
        Mangosteen
      </li>
      <li
        aria-posinset="38"
        aria-selected="false"
        aria-setsize="64"
        class="usa-combo-box__list-option"
        data-value="marionberry"
        id="fruit--option-37"
        role="option"
        tabindex="-1"
      >
        Marionberry
      </li>
      <li
        aria-posinset="39"
        aria-selected="false"
        aria-setsize="64"
        class="usa-combo-box__list-option"
        data-value="nectarine"
        id="fruit--option-38"
        role="option"
        tabindex="-1"
      >
        Nectarine
      </li>
      <li
        aria-posinset="40"
        aria-selected="false"
        aria-setsize="64"
        class="usa-combo-box__list-option"
        data-value="orange"
        id="fruit--option-39"
        role="option"
        tabindex="-1"
      >
        Orange
      </li>
      <li
        aria-posinset="41"
        aria-selected="false"
        aria-setsize="64"
        class="usa-combo-box__list-option"
        data-value="papaya"
        id="fruit--option-40"
        role="option"
        tabindex="-1"
      >
        Papaya
      </li>
      <li
        aria-posinset="42"
        aria-selected="false"
        aria-setsize="64"
        class="usa-combo-box__list-option"
        data-value="passionfruit"
        id="fruit--option-41"
        role="option"
        tabindex="-1"
      >
        Passionfruit
      </li>
      <li
        aria-posinset="43"
        aria-selected="false"
        aria-setsize="64"
        class="usa-combo-box__list-option"
        data-value="peach"
        id="fruit--option-42"
        role="option"
        tabindex="-1"
      >
        Peach
      </li>
      <li
        aria-posinset="44"
        aria-selected="false"
        aria-setsize="64"
        class="usa-combo-box__list-option"
        data-value="pear"
        id="fruit--option-43"
        role="option"
        tabindex="-1"
      >
        Pear
      </li>
      <li
        aria-posinset="45"
        aria-selected="false"
        aria-setsize="64"
        class="usa-combo-box__list-option"
        data-value="persimmon"
        id="fruit--option-44"
        role="option"
        tabindex="-1"
      >
        Persimmon
      </li>
      <li
        aria-posinset="46"
        aria-selected="false"
        aria-setsize="64"
        class="usa-combo-box__list-option"
        data-value="plantain"
        id="fruit--option-45"
        role="option"
        tabindex="-1"
      >
        Plantain
      </li>
      <li
        aria-posinset="47"
        aria-selected="false"
        aria-setsize="64"
        class="usa-combo-box__list-option"
        data-value="plum"
        id="fruit--option-46"
        role="option"
        tabindex="-1"
      >
        Plum
      </li>
      <li
        aria-posinset="48"
        aria-selected="false"
        aria-setsize="64"
        class="usa-combo-box__list-option"
        data-value="pineapple"
        id="fruit--option-47"
        role="option"
        tabindex="-1"
      >
        Pineapple
      </li>
      <li
        aria-posinset="49"
        aria-selected="false"
        aria-setsize="64"
        class="usa-combo-box__list-option"
        data-value="pluot"
        id="fruit--option-48"
        role="option"
        tabindex="-1"
      >
        Pluot
      </li>
      <li
        aria-posinset="50"
        aria-selected="false"
        aria-setsize="64"
        class="usa-combo-box__list-option"
        data-value="pomegranate"
        id="fruit--option-49"
        role="option"
        tabindex="-1"
      >
        Pomegranate
      </li>
      <li
        aria-posinset="51"
        aria-selected="false"
        aria-setsize="64"
        class="usa-combo-box__list-option"
        data-value="pomelo"
        id="fruit--option-50"
        role="option"
        tabindex="-1"
      >
        Pomelo
      </li>
      <li
        aria-posinset="52"
        aria-selected="false"
        aria-setsize="64"
        class="usa-combo-box__list-option"
        data-value="quince"
        id="fruit--option-51"
        role="option"
        tabindex="-1"
      >
        Quince
      </li>
      <li
        aria-posinset="53"
        aria-selected="false"
        aria-setsize="64"
        class="usa-combo-box__list-option"
        data-value="raspberry"
        id="fruit--option-52"
        role="option"
        tabindex="-1"
      >
        Raspberry
      </li>
      <li
        aria-posinset="54"
        aria-selected="false"
        aria-setsize="64"
        class="usa-combo-box__list-option"
        data-value="rambutan"
        id="fruit--option-53"
        role="option"
        tabindex="-1"
      >
        Rambutan
      </li>
      <li
        aria-posinset="55"
        aria-selected="false"
        aria-setsize="64"
        class="usa-combo-box__list-option"
        data-value="soursop"
        id="fruit--option-54"
        role="option"
        tabindex="-1"
      >
        Soursop
      </li>
      <li
        aria-posinset="56"
        aria-selected="false"
        aria-setsize="64"
        class="usa-combo-box__list-option"
        data-value="starfruit"
        id="fruit--option-55"
        role="option"
        tabindex="-1"
      >
        Starfruit
      </li>
      <li
        aria-posinset="57"
        aria-selected="false"
        aria-setsize="64"
        class="usa-combo-box__list-option"
        data-value="strawberry"
        id="fruit--option-56"
        role="option"
        tabindex="-1"
      >
        Strawberry
      </li>
      <li
        aria-posinset="58"
        aria-selected="false"
        aria-setsize="64"
        class="usa-combo-box__list-option"
        data-value="tamarind"
        id="fruit--option-57"
        role="option"
        tabindex="-1"
      >
        Tamarind
      </li>
      <li
        aria-posinset="59"
        aria-selected="false"
        aria-setsize="64"
        class="usa-combo-box__list-option"
        data-value="tangelo"
        id="fruit--option-58"
        role="option"
        tabindex="-1"
      >
        Tangelo
      </li>
      <li
        aria-posinset="60"
        aria-selected="false"
        aria-setsize="64"
        class="usa-combo-box__list-option"
        data-value="tangerine"
        id="fruit--option-59"
        role="option"
        tabindex="-1"
      >
        Tangerine
      </li>
      <li
        aria-posinset="61"
        aria-selected="false"
        aria-setsize="64"
        class="usa-combo-box__list-option"
        data-value="ugli fruit"
        id="fruit--option-60"
        role="option"
        tabindex="-1"
      >
        Ugli fruit
      </li>
      <li
        aria-posinset="62"
        aria-selected="false"
        aria-setsize="64"
        class="usa-combo-box__list-option"
        data-value="watermelon"
        id="fruit--option-61"
        role="option"
        tabindex="-1"
      >
        Watermelon
      </li>
      <li
        aria-posinset="63"
        aria-selected="false"
        aria-setsize="64"
        class="usa-combo-box__list-option"
        data-value="white currant"
        id="fruit--option-62"
        role="option"
        tabindex="-1"
      >
        White currant
      </li>
      <li
        aria-posinset="64"
        aria-selected="false"
        aria-setsize="64"
        class="usa-combo-box__list-option"
        data-value="yuzu"
        id="fruit--option-63"
        role="option"
        tabindex="-1"
      >
        Yuzu
      </li>
    </ul>
    <div
      class="usa-combo-box__status usa-sr-only"
      role="status"
    />
    <span
      class="usa-sr-only"
      id="fruit--assistiveHint"
    >
      When autocomplete results are available use up and down arrows to review and enter to select. Touch device users, explore by touch or with swipe gestures.
    </span>
  </div>
</div>
`;
